GVFS/GVFS.Common/Git/GitSsl.cs

```diff
@@ -0,0 +1,171 @@
+            switch (numberOfCertificates)
+            {
+                case 0:
+                    loggingFunction = tracer.RelatedError;
+                    break;
+                case 1:
+                    loggingFunction = tracer.RelatedInfo;
+                    break;
+                default:
+                    loggingFunction = tracer.RelatedWarning;
+                    break;
+            }
+
+            loggingFunction(
+                metadata,
+                string.Format(
+                    messageTemplate,
+                    numberOfCertificates,
+                    string.Join(
+                        Environment.NewLine,
+                        certificates.Select(x => x.Subject))));
+        }
+
+        private X509Certificate2 GetCertificateFromFile(ITracer tracer, EventMetadata metadata, string certificatePassword, bool onlyLoadValidCertificateFromStore)
+        {
+            if (File.Exists(this.CertificatePathOrSubjectCommonName))
+            {
+                try
+                {
+                    X509Certificate2 cert = new X509Certificate2(this.CertificatePathOrSubjectCommonName, certificatePassword);
+                    if (onlyLoadValidCertificateFromStore && cert != null && !cert.Verify())
+                    {
+                        tracer.RelatedWarning(metadata, "Certficate was found, but is invalid.");
+                        return null;
+                    }
+
+                    return cert;
+                }
+                catch (CryptographicException cryptEx)
+                {
+                    metadata.Add("Exception", cryptEx);
+                    tracer.RelatedError(metadata, "Error, while loading certificate from disk");
+                    return null;
+                }
+            }
+
+            return null;
+        }
+
+        private X509Certificate2 GetCertificateFromStore(ITracer tracer, EventMetadata metadata, bool onlyLoadValidCertificateFromStore)
+        {
+            X509Store store = null;
+            try
+            {
+                store = new X509Store();
+                store.Open(OpenFlags.OpenExistingOnly | OpenFlags.ReadOnly);
+
+                X509Certificate2Collection findResults = store.Certificates.Find(X509FindType.FindBySubjectName, this.CertificatePathOrSubjectCommonName, onlyLoadValidCertificateFromStore);
+                if (findResults?.Count > 0)
+                {
+                    LogCertificateCounts(tracer, metadata, findResults.OfType<X509Certificate2>(), "Found {0} certificates by provided name. Matching DNs: {1}");
+
+                    X509Certificate2[] certsWithMatchingCns = findResults
+                        .OfType<X509Certificate2>()
+                        .Where(x => x.HasPrivateKey && Regex.IsMatch(x.Subject, string.Format("(^|,\\s?)CN={0}(,|$)", this.CertificatePathOrSubjectCommonName))) // We only want certificates, that have private keys, as we need them. We also want a complete CN match

```

Comment:
Not a fan of regex.  They are hard to understand and maintain.  Can we just use
`x.Subject.IndexOf($"CN={this.CertificatePathOrSubjectCommonName}") >= 0` ?

Reply:
They are hard to understand, if you don't know them :)  I agree though, that they're not really readable, but they allow us to achieve precise string matching.
Your suggested code would also match "CN={subjectName}AndSomethingElse" as well as "DCN={subjectName}" - so not an exact match.

As always, this is your code and your standards, so if you're fine with imprecise matching, I can implement your suggested changes.

Reply:
We can keep it.  It is pretty simple and should not be a performance concern in this code path.